core/src/main/java/io/questdb/griffin/engine/groupby/TimestampSampler.java

```diff
@@ -28,11 +28,17 @@
 
 public interface TimestampSampler extends Sinkable {
 
+    long getApproxBucketSize();
+
     default long getBucketSize() {
         throw new UnsupportedOperationException();
     }
 
-    long nextTimestamp(long timestamp);
+    default long nextTimestamp(long timestamp) {
+        return nextTimestamp(timestamp, 1, Long.MAX_VALUE);
+    }
+
+    long nextTimestamp(long timestamp, int numSteps, long maxTimestamp);
```

Comment:
@mtopolnik we should remove `maxTimestamp` from this method's arguments and keep checking it in the iterator. That's because the timestamp has to be aligned at the bucket boundary and implementations of this methods don't check that. So, it's less error-prone to keep it external.

Reply:
Resolved in 1f8d23d